### ISSUE TITLE
use { } for tuple types?

### ISSUE THREAD

**Issue by JeffBezanson** [JuliaLang/julia#8470]
@jakebolewski had this very good idea today.

Our tuple types are currently quite ugly since they are not distinguished from tuple values. This causes various problems in the system, such as whether to treat `()` as a type. We also need to specialize tuple representations (e.g. storing `(Float64,Float64)` as compactly as a `Complex128`), which means tuple types will need to contain extra layout information that tuple values do not have. This would also eliminate the need for a fancy predicate to test whether some tuple is a type.

It would be quite easy to use `Union{...}` for union types, and then all types would be constructed with curly braces :+1:.

This change also makes it easy to use `(x, y...)` for splatting in tuple construction (#4869), and `{Int...}` for varargs types.

All current uses of `{ }` can be replaced with `Any[ ]` or `(Any=>Any)[ ]`, so that syntax is effectively up for grabs. The wastefulness of the current `{ }` has been discussed elsewhere. I admit this proposal feels slightly wasteful as well, but it could be worth it for solving various problems in the type system and making type syntax highly consistent. On the whole, I am in favor of this idea.

**Comment by jiahao** [comment]
Closed by #10380

**Comment by jakebolewski** [comment]
I still hold out hope that this will be the final syntax.

**Comment by JeffBezanson** [comment]
I agree. I'm 100% in favor of this. I'm tired of typing "Tuple", and hopefully soon everybody else will be too.

**Comment by tkelman** [comment]
Based on the deprecation policy, we should really leave `{}` as its former deprecated meaning for at least one release. So I don't think this syntax change needs to be resolved until early in 0.5-dev.

**Comment by JeffBezanson** [comment]
In that case, we might want to continue allowing tuples of types in various places. Writing `code_typed(foo, Tuple{ ... })` gets old.

**Comment by StefanKarpinski** [comment]
Can't you do that just by writing methods that convert tuples to tuple types?

**Comment by JeffBezanson** [comment]
Yes, the only downside is that it creates some confusion about what is what.

**Comment by mauro3** [comment]
What about the `Union{...}` syntax mentioned above?  Now would be a prefect time to change it.

**Comment by JeffBezanson** [comment]
Yes I'd like to change that soon.

**Comment by lindahua** [comment]
I agree that it is now the perfect time to make such changes.

**Comment by tkelman** [comment]
A few months ago would've been the perfect time... let's not try adding too much additional breakage on top of what we've already got, so package authors have a chance to fix things with a smaller number of changes at a time.

Deprecations for tuples of types in a few places would be very helpful, many of the currently broken packages could probably be fixed with a small number of band-aid deprecations.

**Comment by JeffBezanson** [comment]
Call overloading allows the Union change to be pretty gentle. We can easily either deprecate it, or just support both syntaxes for a while.

**Comment by andrewcooke** [comment]
not sure if this is the right place to ask, but it seems odd to [everyone here](https://groups.google.com/forum/#!topic/julia-users/zzeUntRbQ34) that Tuple{...}(...) takes a *tuple* as an argument.  it seems more natural (more analogous with the implicit immutable type) to have `Tuple{Int,Int}(1,2)` than `Tuple{Int,Int}((1,2))`.  apart from the whole chicken and egg thing that's going on there.

**Comment by JeffBezanson** [comment]
What's happening is that there are no methods of `call` specifically for tuple types, so it falls back to calling `convert`, and you can only `convert` a tuple to a tuple type. `call` methods could certainly be added.

Another issue is that you can't directly "pick" which tuple type to get: `convert(Tuple{Integer}, (1,))` will actually return a `Tuple{Int}`.

**Comment by JeffBezanson** [comment]
Bump. `{ }` has been deprecated for a full release; I propose we do this.

**Comment by tkelman** [comment]
Let's remove the old deprecated meaning now, then give it a week or two before introducing a new meaning for it. I don't particularly mind typing `Tuple{}` myself.

**Comment by mauro3** [comment]
+1 to using `{}`, +1 to have a gap.

**Comment by nalimilan** [comment]
+1 to both too

**Comment by StefanKarpinski** [comment]
I'm on board with removing the deprecated meaning but I'm not entirely convinced that tuple types are the most profitable use of this syntax.

**Comment by KristofferC** [comment]
+1 @StefanKarpinski I also think we can find something better for`{}`. It could perhaps find use in some of the array changes coming up.

**Comment by mbauman** [comment]
Random brainstorm: what about this syntax for tuple types: `(::Int, ::Int)`?  As far as I can tell, it is available.  It parses but is a syntax error.  And it very nicely mirrors argument list syntax — it could even support varargs `(::Int, ::Int...)`.  I think it also makes the covariant behavior clearer, since it's just like an argument list.

```jl
methods(convert, (::Any, ::AbstractString))
function foo()::(::Int, ::String)
head(tup::(::Any, ::Any...)) = tup[1]
```

Downsides: Lots of line noise in `function f(::A, ::B)::(::T, ::S, ::R...)`.  And it's a little strange that `(a::Int, b::String)` is a tuple with type-asserted components but `(::Int, ::String)` is `Tuple{Int, String}`.

**Comment by StefanKarpinski** [comment]
I like that but yeah, it's a bit verbose.

**Comment by mbauman** [comment]
It's also problematic in that it cannot express `Tuple{}`.

**Comment by davidanthoff** [comment]
I write *a lot* of ``Tuple{...}`` when I'm wearing my package developer hat. But I hardly ever need it when I work on actual science code, so I'm also -1 to use the ``{ }`` syntax for tuple types. I think package devs can suffer through the more verbose existing syntax and ``{ }`` should be used for something that normal users would use more often.

My preference would be that ``{ }`` is used for some future base support of named tuples/anonymous types, like in C#. Essentially I would like it to be a nicer syntax for the functionality that is currently provided in [NamedTuples.jl](https://github.com/blackrock/NamedTuples.jl). My motivation is that it would make my hopefully soon to be released [Query.jl](https://github.com/davidanthoff/Query.jl) package much nicer :)

**Comment by JeffBezanson** [comment]
We can probably use the syntax `(a=1, b=2, ...)` for anonymous structs / named tuples, so `{ }` is not needed for that. In fact at that point using `{ }` for tuple types *and* struct types becomes more compelling, e.g. `typeof(("", a=1)) == {String; a::Int}`.

**Comment by davidanthoff** [comment]
Ah, even better on the named tuples side, that would be perfect syntax!

I'm still not entirely convinced that it wouldn't be better to use ``{}`` for something that average users will find useful, but given that I don't even have an idea, I don't really feel strongly about it ;)

**Comment by JeffBezanson** [comment]
At this point I'll link to #16580.

One interesting thing about NamedTuples is that I've found it useful that they are ordered. However, struct/record types are usually not, so we'll have to think about how to reconcile this if we want to subsume both NamedTuples and features like keyword arguments.

**Comment by davidanthoff** [comment]
Actually, let me retract my enthusiasm for ``(a=1, b=2, ...)`` as the syntax for named tuples. One thing that would really help [Query.jl](https://github.com/davidanthoff/Query.jl) is syntax that automatically names members of the named tuple with the name of the variable/member that is passed in. For example (assuming ``{}`` is the named tuple syntax): ``{a=1,b,i.c}`` would create a named tuple with members, ``a``, ``b`` and ``c`` (where ``b`` is a variable from the surrounding scope and ``i`` is a variable of some composite type). That kind of syntax makes LINQ really easy to use, so it would be great for Query.jl if julia named tuples would support similar syntax.

**Comment by JeffBezanson** [comment]
We might be able to use the syntax `(a=1; b, i.c)`, or `(; a=1, b, i.c)` for that. Currently (for keyword arguments) this means `b` is expected to evaluate to a `symbol=>value` pair, but that isn't essential --- we could require writing `a=>b` directly for passing computed keywords.

Another consideration is that since functions have both positional and named arguments, we also need some data structure that has both. This would naturally be written as `(a, b; c=1, d=2)`, like an argument list. In that situation it would be a bit redundant for `{ }` to be syntax for named-only tuples.

In a context where `b` implicitly expands to `:b => b`, what does a more general expression like `f(x)` expand to? I would frown on built-in language syntax that doesn't accept arbitrary expressions. The "everything just gets evaluated" semantics of lisp-family languages is a large part of what makes them powerful and intuitive.

**Comment by eschnett** [comment]
The syntax `(1; a=2)` already has a well-defined meaning; it is interpreted as a sequence of two statements. The parser currently has "difficulties" telling parenthesized statement sequences and tuples apart, in the sense that my intuition and its interpretation often differ. Making the tuple syntax more flexible will make this even more problematic (for me).

One way out would be to require that statement lists need to be enclosed in `begin ... end` instead of parentheses. Another way out would be to use `{ ... }` as constructor for tuples. The latter would also avoid the pesky trailing comma for single-element tuples.

**Comment by StefanKarpinski** [comment]
... and that, kids, is how Julia ended up using `{ }` to delimit blocks.

**Comment by JeffBezanson** [comment]
I do think the `(a; b)` statement syntax is a bit marginal and we might be better off deprecating it and using `begin .. end`.

However, the trailing comma trick still works with semicolon:

```
> (julia-parse "(1,; 2)")
(tuple (parameters 2) 1)

> (julia-parse "(1,; a=2)")
(tuple (parameters (kw a 2)) 1)
```

I don't think `(1, 2)` as tuple syntax can be avoided easily. If `a, b = 1, 2` works, then `a, b = (1, 2)` should work, so we'd be left with both `(1, 2)` and `{1, 2}` as tuple syntax.

**Comment by eschnett** [comment]
We could reserve the curlies for something really cool, maybe parallelism related. It could be part of a simple syntax for a future, or a reference to a remote object, or a task, etc.

**Comment by davidanthoff** [comment]
At least for [Query](https://github.com/davidanthoff/Query.jl), I'm not a fan of something like ``(; a=1, b, i.c)``. It would mean that even the simplest select statement that selects one column from a table would look something like ``@select (;i.a)``, which is aesthetically just not very pleasing... ``@select {i.a}`` seems much more natural (and I just implemented that syntax in [Query](https://github.com/davidanthoff/Query.jl) as an experimental feature on master).

Now, I understand that [Query](https://github.com/davidanthoff/Query.jl)'s needs are not the decisive factor here ;), but I would probably think that there will more users out there that need something like LINQ (which [Query](https://github.com/davidanthoff/Query.jl) is) than a more convenient tuple type syntax.

> In a context where b implicitly expands to :b => b, what does a more general expression like f(x) expand to?

Something like ``{f(x)}`` would be an error (it is in C#), you would have to write e.g. ``{x=f(x)}`` to construct a named tuple with one element named ``x``.

The more I think about named tuples vs what is needed for function arguments, the less clear it is to me how these two situations could use the same underlying machinery... For named tuples for something like [Query](https://github.com/davidanthoff/Query.jl), the order of the names matters a lot, i.e. ``{a=1, b=2}`` should not be equal to ``{b=2,a=1}``, but for function keyword arguments that of course is not true, right?

**Comment by JeffBezanson** [comment]
This is still controversial; moving to 1.0.

**Comment by JeffBezanson** [comment]
For 1.0, we should change the parsing to not use the expression head `cell1d` (lol), then the rest of this will be non-breaking and this can be moved out of 1.0.

**Comment by JeffBezanson** [comment]
Parser change done.

**Comment by mauro3** [comment]
I don't know whether it is still on the table to use `{}` for `Tuple{}`.  But here one argument against it: in `where` constructs, which already use bare `{}` it would be confusing to have another set of bare `{}`.
```julia
Tuple{X,Y} where X<:Tuple{T,Y} where {Y,T}
# would become
{X,Y} where X<:{T,Y} where {Y,T}
```

**Comment by oscardssmith** [comment]
I would like to +1 {} for sets and dicts. Besides compatibility with python, it goes two very common structures one very convenient syntax

**Comment by timholy** [comment]
Another option is a lisp-like representation of tuples (as a linked list). There's one coming (called `TupleLL`) as part of #23692.

**Comment by davidanthoff** [comment]
> Another option is a lisp-like representation of tuples (as a linked list)

Do we expect that to be widely used by a large fraction of the user base? In my mind that should be an important factor in the decision what to do with ``{}``,

**Comment by timholy** [comment]
Unlikely but possible. Currently "lispy-tuple" programming with ordinary tuples is `O(N^2)` on the compiler, but if we introduced an efficient conversion between ordinary tuples and linked-list tuples then we could make it `O(N)`. If that happened, then they might end up being used relatively widely.

**Comment by sighoya** [comment]
Another problem with {} as Tuple type is the inclusion of values inside the curly brackets:
```julia
Array{Int64,2}
```

which is maybe a problem for future syntax:
```julia
struct A{S,T=S*3} ... end
```

**Comment by StefanKarpinski** [comment]
I really think we should let 1.0 come out, settle in and see where there's real syntactic pain. Having some good syntax available to address pain points that emerge is a good thing. It's hard to predict what the real issues will be and which apparent problems evaporate as people work around them.

**Comment by timholy** [comment]
Agreed, I was definitely not suggesting this for now (that would be crazy), simply planting the seed for some future discussion. Months or years later It's easy to forget these thoughts.

**Comment by ajkeller34** [comment]
> I really think we should let 1.0 come out, settle in and see where there's real syntactic pain.

There's some syntactic pain in conflating indexing by integer offsets and getting keys when both could be considered useful operations on a data structure, for example in AxisArrays but not exclusively there. Curly braces could lower to `get` (which could be renamed `getkey` for clarity, because what am I getting exactly?). See https://github.com/JuliaLang/julia/pull/25057#issuecomment-358580603

**Comment by StefanKarpinski** [comment]
The is now getproperty which addresses some of that. The only syntax that’s available here is braces not used for indexing.

**Comment by RaulDurand** [comment]
Sorry for awakening this old topic. I was introducing the language to a C++ programmer and he was not very attracted to the use of begin-end delimiters.  I was wondering if, at some point, the use of {} as the block delimiter had been discussed.

**Comment by StefanKarpinski** [comment]
Lol

**Comment by haberdashPI** [comment]
for more context @RaulDurand: https://discourse.julialang.org/t/psa-julia-is-not-at-that-stage-of-development-anymore/44872
